databricks_cli/stack/api.py

```diff
@@ -0,0 +1,340 @@
+        :return: None
+        """
+        with open(path, 'w+') as f:
+            json.dump(data, f, indent=2, sort_keys=True, default=self._json_type_handler)
+
+    def _generate_stack_status_path(self, stack_path):
+        """
+        Given a path to the stack configuration template JSON file, generates a path to where the
+        deployment status JSON will be stored after successful deployment of the stack.
+
+        :param stack_path: Path to the stack config template JSON file
+        :return: The path to the stack status file.
+
+        >>> self._generate_stack_status_path('./stack.json')
+        './stack.deployed.json'
+        """
+        stack_path_split = stack_path.split('.')
+        stack_path_split.insert(-1, STACK_STATUS_INSERT)
+        return '.'.join(stack_path_split)
+
+    def _get_previous_resource_deploy_map(self, stack_status):
+        """
+        Returns a dictionary that maps a resource's (id, service) to the resource's metadata
+        from the last deployment
+        """
+        return {(resource[RESOURCE_ID], resource[RESOURCE_SERVICE]): resource
+                for resource in stack_status[STACK_DEPLOYED]}
+
+    def put_job(self, job_settings):
+        """
+        Given settings of the job in job_settings, create a new job. For purposes of idempotency
+        and to reduce leaked resources in alpha versions of stack deployment, if a job exists
+        with the same name, that job will be updated. If multiple jobs are found with the same name,
+        the deployment will abort.
+
+        :param job_settings:
+        :return: job_id, Physical ID of job on Databricks server.
+        """
+        if 'name' not in job_settings:
+            raise StackError("Please supply 'name' in job resource 'resource_properties'")
+        job_name = job_settings['name']
+        jobs_same_name = self.jobs_client._list_jobs_by_name(job_name)
+        if len(jobs_same_name) > 1:
+            raise StackError("Multiple jobs with the same name '%s' already exist, aborting"
+                             " stack deployment" % job_name)
+        elif len(jobs_same_name) == 1:
+            existing_job = jobs_same_name[0]
+            creator_name = existing_job['creator_user_name']
+            timestamp = existing_job['created_time'] / MS_SEC  # Convert to readable date.
+            date_created = datetime.fromtimestamp(timestamp).strftime('%Y-%m-%d %H:%M:%S')
+            click.echo("Warning: Job exists with same name '%s' created by %s on %s. Job will "
+                       "be overwritten" % (job_name, creator_name, date_created))
+            self.update_job(job_settings, existing_job['job_id'])
+            return existing_job['job_id']
+        else:
+            click.echo("Creating new job")
+            job_id = self.jobs_client.create_job(job_settings)['job_id']
+            return job_id
+
+    def update_job(self, job_settings, job_id):
```

Comment:
This function is too thin, we can just use `self.jobs_client.reset_job` directly in 
https://github.com/databricks/databricks-cli/pull/137/files#diff-adef28b314958b556e396054d0efb708R151

Reply:
Do you think there may be some more important logics in the function in the future though? Say if there were a dry-run mode?

Reply:
I see it is also used https://github.com/databricks/databricks-cli/pull/137/files#diff-adef28b314958b556e396054d0efb708R187
It is okey to leave it this way and let `deploy_job` to call `update_job`, how every, `put_job` should not call `update_job`. How about keep the function but use `reset_job` API in `put_job` directly?